### ISSUE TITLE
Unable to compile a truly static binary

### ISSUE THREAD

**Issue by morr0ne** [sunfishcode/origin#23]
I am trying to compile a static binary but I can't seem to get anything to work. Below is a minimal example program explainng the issue.
<details>
<summary>Cargo.toml</summary>

```toml
[package]
name = "origin-static-example"
version = "0.1.0"
edition = "2021"

[dependencies]
origin = { version = "0.12.0", default-features = false, features = [
    "origin-program",
    "origin-start",
] }
```

</details>

<details>
<summary>src/main.rs</summary>

```rust
#![no_std]
#![no_main]
#![allow(internal_features)]
#![feature(lang_items)]
#![feature(core_intrinsics)]

use origin::program::*;

#[panic_handler]
fn panic(_panic: &core::panic::PanicInfo<'_>) -> ! {
    loop {} // Loop forever to ensure we are not causing the segfault ourselves
}

#[lang = "eh_personality"]
extern "C" fn eh_personality() {}

#[no_mangle]
extern "C" fn main(_argc: i32, _argv: *const *const u8) -> i32 {
    exit_immediately(0);  // Exit immediately to ensure the segfault isn't cause by any dtor being called
}
```

</details>

When building with target x86_64-unknown-linux-gnu using the following command
```bash
> cargo +nightly run --release --target x86_64-unknown-linux-gnu
```
Results in a *statically linked* binary, at least according to `ldd`
```bash
> ldd target/x86_64-unknown-linux-gnu/release/origin-static-example
statically linked
```
Unfortunately that is a lie, sort of. While the binary doesn't actually link to anything it still requires a dynamic loader to run, I found this out when trying to run it on a system without dynamic linking support and got a "not found" error. This can also be confirmed by using readelf
```bash
> readelf -l target/x86_64-unknown-linux-gnu/release/origin-static-example
Elf file type is DYN (Position-Independent Executable file)
Entry point 0x1380
There are 11 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000000040 0x0000000000000040
                 0x0000000000000268 0x0000000000000268  R      0x8
  INTERP         0x00000000000002a8 0x00000000000002a8 0x00000000000002a8
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000670 0x0000000000000670  R      0x1000
  LOAD           0x0000000000001000 0x0000000000001000 0x0000000000001000
                 0x0000000000001733 0x0000000000001733  R E    0x1000
  LOAD           0x0000000000003000 0x0000000000003000 0x0000000000003000
                 0x0000000000000a38 0x0000000000000a38  R      0x1000
  LOAD           0x0000000000003d20 0x0000000000004d20 0x0000000000004d20
                 0x00000000000002e8 0x0000000000000320  RW     0x1000
  DYNAMIC        0x0000000000003e48 0x0000000000004e48 0x0000000000004e48
                 0x0000000000000110 0x0000000000000110  RW     0x8
  NOTE           0x00000000000002c4 0x00000000000002c4 0x00000000000002c4
                 0x0000000000000024 0x0000000000000024  R      0x4
  GNU_EH_FRAME   0x00000000000034d4 0x00000000000034d4 0x00000000000034d4
                 0x00000000000000fc 0x00000000000000fc  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  GNU_RELRO      0x0000000000003d20 0x0000000000004d20 0x0000000000004d20
                 0x00000000000002e0 0x00000000000002e0  R      0x1

 Section to Segment mapping:
  Segment Sections...
   00
   01     .interp
   02     .interp .note.gnu.build-id .gnu.hash .dynsym .dynstr .rela.dyn
   03     .text
   04     .rodata .eh_frame_hdr .eh_frame
   05     .data.rel.ro .dynamic .got .data .bss
   06     .dynamic
   07     .note.gnu.build-id
   08     .eh_frame_hdr
   09
   10     .data.rel.ro .dynamic .go
```
As shown in the output, the executale is actually requesting `lib64/ld-linux-x86-64.so.2` as the interpreter.
The first thing I tried is off course to add `target-feature=+crt-static` to the RUSTFLAGS. This does result in a completely static binary, sadly this binary is somehow broken as it segfaults immediately when run
```bash
> ./target/x86_64-unknown-linux-gnu/release/origin-static-example
[1]    107995 segmentation fault (core dumped)  ./target/x86_64-unknown-linux-gnu/release/origin-static-example
```
I also tried building with musl, I did use [cross](https://github.com/cross-rs/cross) since I don't have a better way to test this but it shouldn't make a difference since it's pretty much building using a preconfigured container:
```bash
>cross +nightly build --release --target x86_64-unknown-linux-musl
```

Unfortunately this gives me the following linking error:
```bash
error: linking with `x86_64-linux-musl-gcc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/rust/lib/rustlib/x86_64-unknown-linux-gnu/bin:/rust/lib/rustlib/x86_64-unknown-linux-gnu/bin/self-contained:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/rust/bin" VSLANG="1033" "x86_64-linux-musl-gcc" "-m64" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/rcrt1.o" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crti.o" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtbeginS.o" "/tmp/rustcvUYUvi/symbols.o" "/target/x86_64-unknown-linux-musl/release/deps/origin_static_example-5f75919837218881.origin_static_example.bda4ed8c7e9b5deb-cgu.0.rcgu.o" "-Wl,--as-needed" "-L" "/target/x86_64-unknown-linux-musl/release/deps" "-L" "/target/release/deps" "-L" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib" "-Wl,-Bstatic" "/target/x86_64-unknown-linux-musl/release/deps/liborigin-ddd2949b7032c854.rlib" "/target/x86_64-unknown-linux-musl/release/deps/librustix-98fdeda1a8ab2e69.rlib" "/target/x86_64-unknown-linux-musl/release/deps/libbitflags-4516e257f674d3d7.rlib" "/target/x86_64-unknown-linux-musl/release/deps/liblinux_raw_sys-6a3c212ffded53d3.rlib" "/target/x86_64-unknown-linux-musl/release/deps/libunwinding-6c2b838d90cfef19.rlib" "/target/x86_64-unknown-linux-musl/release/deps/libgimli-101c23ae2dc42dcd.rlib" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/librustc_std_workspace_core-e251179636a4eb0c.rlib" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/libcore-d01acce508fabf16.rlib" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/libcompiler_builtins-7ab2269e6a08c0c8.rlib" "-Wl,-Bdynamic" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-nostartfiles" "-L" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib" "-L" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained" "-o" "/target/x86_64-unknown-linux-musl/release/deps/origin_static_example-5f75919837218881" "-Wl,--gc-sections" "-static-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-nostartfiles" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtendS.o" "/rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/crtn.o"
  = note: /usr/local/bin/../lib/gcc/x86_64-linux-musl/9.2.0/../../../../x86_64-linux-musl/bin/ld: /rust/lib/rustlib/x86_64-unknown-linux-musl/lib/self-contained/rcrt1.o: in function `__dls2':
          /build/musl-cross-make/build/local/x86_64-linux-musl/obj_musl/../src_musl/crt/rcrt1.c:13: undefined reference to `__libc_start_main'
          collect2: error: ld returned 1 exit status

  = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `origin-static-example` (bin "origin-static-example") due to previous error
```

I am not sure if this is something that can be fixed in origin or if it needs better support from rust. My hunch is that this is probably caused by the configuration of the targets expecting a libc somewhere, if that's the case then the most probable solution (assuming this isn't a logic bug in origin) is to create a custom target that doesn't link to libc instead of relaying on `link-args=-nostartfiles`

**Comment by morr0ne** [comment]
After debuging the program with gdb I figured out the offending cose is [this call to call_ctors](https://github.com/sunfishcode/origin/blob/2edde5becb412bd7dc202f4cf3503eb74153ae74/src/program.rs#L92) in the entry function. Commenting that line out make the code compiled with `target-feature=+crt-static` work without issues. I have not investigated why it causes a segfault but I'm assuming it's because of the way it tries to read .init_array

**Comment by morr0ne** [comment]
So after have bit of digging I determined this is not really an issue with origin implementations but more or less with the fact that it's piggy backing on existing target, that simply wont work cause existing rust target expect some precondition to be met and make assumption about both the binary and the enviroment it will be run on.

The better solution actually goes beyond origin itself: creating a rust only target. Basing it on the existing gnu target I created a spec for a `x86_64-unknown-linux` target that doesn't depend on any enviroment such as gnu.

<details>
<summary>x86_64-unknown-linux.json</summary>

```json
{
  "llvm-target": "x86_64-unknown-linux",
  "target-pointer-width": "64",
  "cpu": "x86-64",
  "data-layout": "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128",
  "is-builtin": false,
  "arch": "x86_64",
  "dynamic-linking": true,
  "max-atomic-width": 64,
  "os": "linux",
  "position-independent-executables": true,
  "static-position-independent-executables": true,
  "target-family": [
    "unix"
  ],
  "linker": "rust-lld",
  "linker-flavor": "ld.lld",
  "has-rpath": true,
  "has-thread-local": true,
  "relro-level": "full",
  "stack-probes": { "kind": "inline" },
  "supported-sanitizers": [
    "address",
    "cfi",
    "leak",
    "memory",
    "thread",
    "safestack"
  ],
  "supported-split-debuginfo": [
    "packed",
    "unpacked",
    "off"
  ],
  "supports-xray": true,
  "crt-static-respected": true,
  "plt-by-default": false
}
```

</details>

The main difference being that it doesn't depend on gnu and it links with lld by default. This also negates the need for `link-args=-nostartfiles`. In fact it will not even compile unless that flag is removed. 

The better long term solution is to open an RFC for the rust language that adds such a target, that however is a long complicated process compared to adding support to origin. My proposal is to officialy support this method of compiling origin programs instead of the old one. The only extra requirement from a user of the library would be to download the target spec and use that in the `--target` flag. This of course requires nightly rust but that was already the case with origin anyway

**Comment by morr0ne** [comment]
Oops, forgot to mention that it also requires to build std for that target using `-Zbuild-std=core`. That also requires for `rust-src` to be installed.

**Comment by sunfishcode** [comment]
The segfault with `-C target-feature=+crt-static` happens because Rust defaults to producing a Position-Independent Executable (PIE), and origin doesn't know how to dynamically relocate itself, so it starts running at a random address but as soon as it makes a call to a function pointer that isn't relocated, it segfaults. This can be worked around by disabling PIE, by adding `-C relocation-model=static` to `RUSTFLAGS`. That allows the origin-start example to produce a fully static executable.

**Comment by morr0ne** [comment]
> The segfault with `-C target-feature=+crt-static` happens because Rust defaults to producing a Position-Independent Executable (PIE), and origin doesn't know how to dynamically relocate itself, so it starts running at a random address but as soon as it makes a call to a function pointer that isn't relocated, it segfaults. This can be worked around by disabling PIE, by also adding `-C relocation-model=static` to `RUSTFLAGS`. That allows the origin-start example to produce a fully static executable.

Using the custom target I created Rust does produce a Position-Independent Executable, origin works fine though

**Comment by morr0ne** [comment]
Okay so apparently my target is getting compiled in `static-pie` mode which somehow creates a binary that's both static and PIE. I'm not sure if that's correct or desired? Searching online gave conflicting results, at this point I'm not sure if there is any reason for an executable to be static and PIE at the same time.

**Comment by sunfishcode** [comment]
I don't know how that works then. Origin itself currently has no knowledge of dynamic relocations, so if the executable is truly running at a random address, something else must be doing those relocations. Or, perhaps you're using a very small example which doesn't happen to call any function pointers loaded from static memory that need to be relocated.

**Comment by morr0ne** [comment]
> I don't know how that works then. Origin itself currently has no knowledge of dynamic relocations, so if the executable is truly running at a random address, something else must be doing those relocations

Truth be told I am not knowledgeable enough about the topic to figure out what is causing the relocations or why, I am learning about this low level stuff along with origin, hearing that is kinda disconcerting




> Or, perhaps you're using a very small example which doesn't happen to call any function pointers loaded from static memory that need to be relocated.

That might be the case but it's very unlikely since the code didn't work with just crt-static and segfaulted. Perhaps lld is smart enough to relocate addresses when building a static-pie executable?

**Comment by morr0ne** [comment]
> If that contains a `RELA` entry, then there are dynamic relocations.

When using my custom target it does indeed have dymanic relocations



> Does your example use any code in `core::fmt`? That's where I see the segfault in the origin-start example. `core::fmt` is calling a function pointer in a vtable, which needs a relocation.

The entire code is the one found in the first issue, it does nothing more than exit with code 0, I even made sure to use exit_immediately to run even less code and didn't use abort anywhere in the code so the segfault cannot be caused by a panic

**Comment by sunfishcode** [comment]
> The entire code is the one found in the first issue, it does nothing more than exit with code 0, I even made sure to use exit_immediately to run even less code and didn't use abort anywhere in the code so the segfault cannot be caused by a panic

My guess then is that this example is just happening to avoid any function pointers that need to be relocated. I'd guess that if you  add realistic application code, you'll hit segfaults.

**Comment by morr0ne** [comment]
> My guess then is that this example is just happening to avoid any function pointers that need to be relocated. I'd guess that if you add realistic application code, you'll hit segfaults.

I suppose that might be the case. What code in origin could even trigger that tho? I am not using any threads, only startup and shutdown

Also maybe I'm not understanding properly but when building a normal dynamically linked application isn't that also a PIE?

**Comment by sunfishcode** [comment]
Dynamic relocations can come up with a rust trait object, or a rust closure, or other language features. There are some in `core`, such as in `core::fmt`. I would be reluctant to make guarantees about code not having any.

Yes, most toolchains default to PIE these days. Normally, ld-linux.so performs the dynamic relocations for the main executable before it starts. That's how things like the origin-start example work right now. So if you want fully static linking, with no ld-linux.so, then the executable needs to do any dynamic relocations for itself.

**Comment by sunfishcode** [comment]
Full static linking is now implemented in #25.
